lib.sh

```diff
@@ -430,11 +430,41 @@ else
 fi
 }
 
-# Let's Encrypt for subdomains
-le_subdomain() {
+#generate certs and auto-configure
+# https://certbot.eff.org/docs/using.html#certbot-command-line-options
+generate_cert() {
+if [[ "no" == $(ask_yes_or_no "Do you need to add the upgrade-insecure-requests(UIR) and Strict-Transport-Security(HSTS) header to every HTTP response?") ]]
+then
+    uir_hsts=""
+fi
+default_le="--rsa-key-size 4096 --renew-by-default --no-eff-email --agree-tos $uir_hsts --server https://acme-v02.api.letsencrypt.org/directory -d $1"
 a2dissite 000-default.conf
 service apache2 reload
-certbot certonly --standalone --pre-hook "service apache2 stop" --post-hook "service apache2 start" --agree-tos --rsa-key-size 4096 -d "$SUBDOMAIN"
+#http-01
+local  standalone="certbot certonly --standalone --pre-hook 'service apache2 stop' --post-hook 'service apache2 start' $default_le"
+#tls-alpn-01
+local  tls_alpn_01="certbot certonly --preferred-challenges tls-alpn-01 $default_le"
+#dns
+local  dns="certbot certonly --manual --manual-public-ip-logging-ok --preferred-challenges dns $default_le"
+local  methods=(standalone dns)
+for f in "${methods[@]}"
+do
+    print_text_in_color "${ICyan}" "try to generate certs and validate them with $f method."
+    eval  ff=$(echo "\$$f")
+    if eval "$ff"
+    then
+        return 0
+    elif [ "$f" != "${methods[@]:(-1)}" ]
+    then
+        print_text_in_color "${ICyan}" "It seems like no certs were generated when trying to validate them with the $f method. We will do more tries."
+        any_key "Press any key to continue..."
+    else
+        print_text_in_color "${ICyan}" "It seems like no certs were generated when trying to validate them with the $f method. We have tried all the methods."
+        print_text_in_color "${IRed}" "Can't generate certs and validate them.  Please check your DNS and try again."
+        any_key "Press any key to continue..."
+        return 1;
```

Comment:
Put the cleanup here maybe?

Reply:
when failure happens in the progress of installing onlyoffice， the script of automatic updating of nextcloud‘s certs will fail because of uninstall of certbot

Reply:
Aah, true. I'm tired now. Thanks!